altair/vegalite/v5/api.py

```diff
@@ -1716,12 +1716,30 @@ def _top_schema_base(  # noqa: ANN202
     """
     Enforces an intersection type w/ `SchemaBase` & `TopLevelMixin` objects.
 
-    Use for instance methods.
+    Use for methods, called from `TopLevelMixin` that are defined in `SchemaBase`.
+
+    Notes
+    -----
+    - The `super` sub-branch is not statically checked *here*.
+        - It would widen the inferred intersection to:
+            - `(<subclass of SchemaBase and TopLevelMixin> | super)`
+        - Both dunder attributes are not in the `super` type stubs
+            - Requiring 2x *# type: ignore[attr-defined]*
+    - However it is required at runtime for any cases that use `super(..., copy)`.
+    - The inferred type **is** used statically **outside** of this function.
     """
-    if isinstance(obj, core.SchemaBase) and isinstance(obj, TopLevelMixin):
+    SchemaBase = core.SchemaBase
```

Comment:
Here I was trying to avoid this issue [runtime-import-in-type-checking-block](https://docs.astral.sh/ruff/rules/runtime-import-in-type-checking-block/)

![image](https://github.com/user-attachments/assets/f724564a-7023-4ced-8bad-a8cfbbf212f4)

To simplify the issue, in `api.py` all uses of `SchemaBase` occur within a type annotation - but all runtime code uses `core.SchemaBase`.
You can see the same pattern for all the imports from `schema.core` - e.g. `core.Binding...`

https://github.com/vega/altair/blob/5831a39edeccdb9761c369073841e6cd46f793ad/altair/vegalite/v5/api.py#L1608-L1635
 

However, since this change https://github.com/vega/altair/pull/3556/files - I think this makes more sense importing from `utils`.

We've already got an import from there, and this is easier to follow to the source:

- `utils.schemapi` -> `vegalite.v5.schema.core` -> `api`
- `utils.schemapi` -> `utils` -> `api`

https://github.com/vega/altair/blob/5831a39edeccdb9761c369073841e6cd46f793ad/altair/vegalite/v5/api.py#L30

Reply:
Fixed in [`b42364b` (#3530)](https://github.com/vega/altair/pull/3530/commits/b42364b7d1edb38ebc6f7321efea0229d2a6cb35)

Also replaced the other `core.SchemaBase` references in `api.py` that are no longer needed

Reply:
Thanks for the explanation. Your new solution makes sense to me.